sway/commands/bind.c

```diff
@@ -649,101 +646,144 @@ void seat_execute_command(struct sway_seat *seat, struct sway_binding *binding)
+		total_combinations *= sets[i]->length;
+	}
+		
+	// Allocate memory for the current_result
+	int result_size = 0;
+	xkb_keycode_t **result = malloc(total_combinations * sizeof(xkb_keycode_t*));
+	xkb_keycode_t *current_product = malloc(n * sizeof(xkb_keycode_t));
+	cartesian_product_helper(sets, n, result, &result_size, current_product, 0);
+	free(current_product);
+
+	return result;
 }
 
+/*
+ * Convert keysyms to keycodes for --to-code. If any keysym does not have at
+ * least 1 keycode: halt conversion and unset BINDING_CODE. Assumes
+ * identify_key() read in binding->keys in keysym format vs keycode format
+ */
 bool translate_binding(struct sway_binding *binding) {
-	if ((binding->flags & BINDING_CODE) == 0) {
-		return true;
-	}
 
-	switch (binding->type) {
-	// a bindsym to translate
-	case BINDING_KEYSYM:
-		binding->syms = binding->keys;
-		binding->keys = create_list();
-		break;
-	// a bindsym to re-translate
-	case BINDING_KEYCODE:
-		list_free_items_and_destroy(binding->keys);
-		binding->keys = create_list();
-		break;
-	default:
+	if (binding->type != BINDING_KEYSYM ||
+		(binding->flags & BINDING_CODE) == 0) {
 		return true;
 	}
 
-	for (int i = 0; i < binding->syms->length; ++i) {
-		xkb_keysym_t *keysym = binding->syms->items[i];
-		struct keycode_matches matches = get_keycode_for_keysym(*keysym);
+	int num_syms = binding->keys->length;
+	list_t **sym2code = malloc(num_syms * sizeof(list_t*)); 
+	// Collect all keycodes for all keysyms
+	for (int i = 0; i < num_syms; i++) {
+		struct keycode_matches matches = {
+			.keysym = (xkb_keysym_t*)binding->keys->items[i],
+			.keycodes = create_list(),
+			.error = false
+		};
 
-		if (matches.count != 1) {
-			sway_log(SWAY_INFO, "Unable to convert keysym %" PRIu32 " into"
-					" a single keycode (found %d matches)",
-					*keysym, matches.count);
+		xkb_keymap_key_for_each(
+				xkb_state_get_keymap(config->keysym_translation_state),
+				add_matching_keycodes, &matches);
+
+		if (matches.error) {
+			sway_log(SWAY_ERROR, "Failed to allocate memory for keycodes while iterating for keysym %" PRIu32, *matches.keysym);
 			goto error;
 		}
 
-		xkb_keycode_t *keycode = malloc(sizeof(xkb_keycode_t));
-		if (!keycode) {
-			sway_log(SWAY_ERROR, "Unable to allocate memory for a keycode");
-			goto error;
+		if (matches.keycodes->length == 0) {
+		   sway_log(SWAY_INFO, "Unable to convert keysym %" PRIu32 " into"
+					  "any keycodes", *matches.keysym);
+		   goto error;
 		}
 
-		*keycode = matches.keycode;
-		list_add(binding->keys, keycode);
+		sym2code[i] = matches.keycodes;
 	}
 
-	list_qsort(binding->keys, key_qsort_cmp);
-	binding->type = BINDING_KEYCODE;
+	// If any keycode maps to more than one keysym, use all combinations.
+	xkb_keycode_t **combos = cartesian_product(sym2code, num_syms);
+
+	for (int i = 0; i < num_syms; i++) {
+		struct sway_binding * copy_binding = malloc(sizeof(struct sway_binding));
+		binding->type = BINDING_KEYCODE;
+		*copy_binding = *binding;
```

Comment:
This isn't safe as it just copies the pointer to strings and lists, which will then be free'd multiple times later.

Reply:
Could we add a check when freeing? Or would it be better to duplicate the command per each binding.

Reply:
It's not impossible - the most likely solution being a ref-counted substructure that can be shared between bindings - but doubt it's worth it over just doing a deep copy to avoid the issue. For strings it's just an strdup, and we presumably don't have the other list set so we can just explicitly NULL it.

Reply:
> which will then be free'd multiple times later.

If we set pointers to null after freeing them, this shouldn't be an issue though right.

Reply:
That would only set them to `NULL` in one copy of the struct, the other copies would still have the pointer set and attempt to free it.